File: typegraph/python/typegraph/graph/typegraph.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

import inspect
from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING, Callable, List, Optional, Union, Any

from typegraph.gen.exports.core import (
    ArtifactResolutionConfig,
    Rate,
    TypegraphInitParams,
)
from typegraph.gen.exports.core import (
    Cors as CoreCors,
)

from typegraph.gen.types import Err
from typegraph.graph.params import Auth, Cors, RawAuth
from typegraph.graph.shared_types import FinalizationResult, TypegraphOutput
from typegraph.policy import Policy, PolicyPerEffect, PolicySpec, get_policy_chain
from typegraph.wit import core, store, wit_utils

if TYPE_CHECKING:
    from typegraph import t


# ExposeItem = Union["t.func", Dict[str, "ExposeItem"]]
ExposeItem = Union["t.func", "t.struct"]


class Typegraph:
    name: str
    dynamic: Optional[bool]
    path: str
    _context: List["Typegraph"] = []
    rate: Optional[Rate]
    cors: Optional[CoreCors]
    prefix: Optional[str]

    def __init__(
        self,
        name: str,
        dynamic: Optional[bool] = None,
        *,
        rate: Optional[Rate] = None,
        cors: Optional[Cors] = None,
        prefix: Optional[str] = None,
    ):
        self.name = name
        self.dynamic = dynamic
        self.path = str(Path(inspect.stack()[2].filename).resolve().parent)

        self.rate = rate

        cors = cors or Cors()
        self.cors = CoreCors(
            allow_origin=cors.allow_origin,
            allow_headers=cors.allow_headers,
            expose_headers=cors.expose_headers,
            allow_methods=cors.allow_methods,
            allow_credentials=cors.allow_credentials,
            max_age_sec=cors.max_age_sec,
        )
        self.prefix = prefix

    @classmethod
    def get_active(cls) -> Optional["Typegraph"]:
        if len(cls._context) == 0:
            raise Exception("No active typegraph")
        return cls._context[-1]

    def __call__(self, **kwargs: ExposeItem):
        self.expose(**kwargs)

    def expose(
        self,
        default_policy: Optional[PolicySpec] = None,
        **kwargs: ExposeItem,
    ):
        res = core.expose(
            store,
            [(k, v.id) for k, v in kwargs.items()],
            default_policy=get_policy_chain(default_policy) if default_policy else None,
        )

        if isinstance(res, Err):
            raise Exception(res.value)


@dataclass
class ApplyFromArg:
    name: Optional[str]


@dataclass
class ApplyFromStatic:
    value: Any


@dataclass
class ApplyFromSecret:
    key: str


@dataclass
class ApplyFromContext:
    key: str


@dataclass
class ApplyFromParent:
    type_name: str


@dataclass
class Graph:
    typegraph: Typegraph

    def expose(
        self,
        default_policy: Optional[Union[Policy, PolicyPerEffect]] = None,
        **kwargs: ExposeItem,
    ):
        self.typegraph.expose(default_policy, **kwargs)

    def inherit(self):
        from typegraph.injection import InheritDef

        return InheritDef()

    def rest(self, graphql: str) -> int:
        res = wit_utils.add_graphql_endpoint(store, graphql)
        if isinstance(res, Err):
            raise Exception(res.value)
        return res.value

    def auth(self, value: Union[Auth, RawAuth]):
        res = (
            wit_utils.add_raw_auth(store, value.json_str)
            if isinstance(value, RawAuth)
            else wit_utils.add_auth(store, value)
        )
        if isinstance(res, Err):
            raise Exception(res.value)
        return res.value

    def ref(self, name: str) -> "t.typedef":
        return gen_ref(name)

    def configure_random_injection(self, seed: int):
        res = core.set_seed(store, seed)
        if isinstance(res, Err):
            raise Exception(res.value)

    def as_arg(self, name: Optional[str] = None):
        return ApplyFromArg(name)

    def set(self, value: Any):
        return ApplyFromStatic(value)

    def from_secret(self, key: str):
        return ApplyFromSecret(key)

    def from_context(self, key: str):
        return ApplyFromContext(key)

    def from_parent(self, type_name: str):
        return ApplyFromParent(type_name)


def typegraph(
    name: Optional[str] = None,
    *,
    dynamic: Optional[bool] = None,
    rate: Optional[Rate] = None,
    cors: Optional[Cors] = None,
    prefix: Optional[str] = None,
) -> Callable[[Callable[[Graph], None]], Callable[[], TypegraphOutput]]:
    def decorator(builder: Callable[[Graph], None]) -> TypegraphOutput:
        actual_name = name
        if name is None:
            import re

            # To kebab case
            actual_name = re.sub("_", "-", builder.__name__)

        tg = Typegraph(
            name=actual_name,
            dynamic=dynamic,
            rate=rate,
            cors=cors,
            prefix=prefix,
        )

        Typegraph._context.append(tg)

        core.init_typegraph(
            store,
            TypegraphInitParams(
                name=tg.name,
                dynamic=tg.dynamic,
                path=tg.path,
                rate=tg.rate,
                cors=tg.cors,
                prefix=tg.prefix,
            ),
        )

        builder(Graph(tg))

        popped = Typegraph._context.pop()
        assert tg == popped

        # config is only known at deploy time
        def serialize_with_artifacts(
            config: ArtifactResolutionConfig,
        ):
            finalization_result = core.finalize_typegraph(store, config)
            if isinstance(finalization_result, Err):
                raise Exception(finalization_result.value)

            tg_json, ref_artifacts = finalization_result.value
            return FinalizationResult(tg_json, ref_artifacts)

        tg_output = TypegraphOutput(name=tg.name, serialize=serialize_with_artifacts)

        from typegraph.graph.tg_manage import Manager

        if Manager.is_run_from_cli():
            manager = Manager(tg_output)
            manager.run()

        return lambda: tg_output

    return decorator


def gen_ref(name: str) -> "t.typedef":
    res = core.refb(store, name, [])
    if isinstance(res, Err):
        raise Exception(res.value)
    from typegraph.t import typedef

    return typedef(res.value)

File: typegraph/deno/sdk/src/typegraph.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import * as t from "./types.ts";
import { core } from "./gen/typegraph_core.js";
import { caller, dirname, fromFileUrl } from "./deps/mod.ts";
import { InjectionValue } from "./utils/type_utils.ts";
import {
  serializeFromParentInjection,
  serializeGenericInjection,
  serializeStaticInjection,
} from "./utils/injection_utils.ts";
import { Auth, Cors as CorsWit, Rate, wit_utils } from "./wit.ts";
import { getPolicyChain } from "./types.ts";
import {
  Artifact,
  SerializeParams,
} from "./gen/interfaces/metatype-typegraph-core.d.ts";
import { Manager } from "./tg_manage.ts";
import { log } from "./io.ts";
import { hasCliEnv } from "./envs/cli.ts";
import process from "node:process";

type Exports = Record<string, t.Func>;

type Cors = Partial<CorsWit>;

interface TypegraphArgs {
  name: string;
  dynamic?: boolean;
  builder: TypegraphBuilder;
  prefix?: string;
  secrets?: Array<string>;
  cors?: Cors;
  rate?: Rate;
  disableAutoSerialization?: boolean;
}

export class ApplyFromArg {
  constructor(public name: string | null, public type: number | null) {}
}

export class ApplyFromStatic {
  constructor(public value: any) {}
}

export class ApplyFromSecret {
  constructor(public key: string) {}
}

export class ApplyFromContext {
  constructor(public key: string | null, public type: number | null) {}
}

export class ApplyFromParent {
  constructor(public typeName: string) {}
}

const InjectionSource = {
  asArg: (name?: string, type?: t.Typedef) =>
    new ApplyFromArg(name ?? null, type?._id ?? null),
  set: (value: any) => new ApplyFromStatic(value),
  fromSecret: (key: string) => new ApplyFromSecret(key),
  fromContext: (key: string, type?: t.Typedef) =>
    new ApplyFromContext(key, type?._id ?? null),
  fromParent: (typeName: string) => new ApplyFromParent(typeName),
} as const;

type InjectionSourceType = typeof InjectionSource;

export interface TypegraphBuilderArgs extends InjectionSourceType {
  expose: (
    exports: Exports,
    defaultPolicy?: t.PolicySpec | Array<t.PolicySpec>
  ) => void;
  inherit: () => InheritDef;
  rest: (graphql: string) => number;
  auth: (value: Auth | RawAuth) => number;
  ref: (name: string) => t.Typedef;
  configureRandomInjection: (params: { seed: number }) => void;
}

export class InheritDef {
  public payload: string | undefined;
  set(value: InjectionValue<unknown>) {
    this.payload = serializeStaticInjection(value);
    return this;
  }

  inject(value: InjectionValue<string>) {
    this.payload = serializeGenericInjection("dynamic", value);
    return this;
  }

  fromContext(value: InjectionValue<string>) {
    this.payload = serializeGenericInjection("context", value);
    return this;
  }

  fromSecret(value: InjectionValue<string>) {
    this.payload = serializeGenericInjection("secret", value);
    return this;
  }

  fromParent(value: InjectionValue<string>) {
    this.payload = serializeFromParentInjection(value);
    return this;
  }

  fromRandom() {
    this.payload = serializeGenericInjection("random", null);
    return this;
  }
}

export type TypegraphBuilder = (g: TypegraphBuilderArgs) => void;

export class RawAuth {
  constructor(readonly jsonStr: string) {}
}

export interface TypegraphOutput {
  serialize: (config: SerializeParams) => TgFinalizationResult;
  name: string;
}

export interface TgFinalizationResult {
  tgJson: string;
  ref_artifacts: Artifact[];
}

let counter = 0;

export async function typegraph(
  name: string,
  builder: TypegraphBuilder
): Promise<TypegraphOutput>;
export async function typegraph(args: TypegraphArgs): Promise<TypegraphOutput>;
export async function typegraph(
  args: Omit<TypegraphArgs, "builder">,
  builder: TypegraphBuilder
): Promise<TypegraphOutput>;
export async function typegraph(
  nameOrArgs: string | TypegraphArgs | Omit<TypegraphArgs, "builder">,
  maybeBuilder?: TypegraphBuilder
): Promise<TypegraphOutput> {
  ++counter;
  const args =
    typeof nameOrArgs === "string" ? { name: nameOrArgs } : nameOrArgs;

  const { name, dynamic, cors, prefix, rate, secrets } = args;
  const builder =
    "builder" in args ? (args.builder as TypegraphBuilder) : maybeBuilder!;

  const file = caller();
  if (!file) {
    throw new Error("Could not determine caller file");
  }
  // node/deno compat tick until MET-236 is landed
  const simpleFile = file.replace(/:[0-9]+$/, "").replace(/^file:\/\//, "");
  const path = dirname(fromFileUrl(`file://${simpleFile}`));

  const defaultCorsFields = {
    allowCredentials: true,
    allowHeaders: [],
    allowMethods: [],
    allowOrigin: [],
    exposeHeaders: [],
    maxAgeSec: undefined,
  } as CorsWit;

  const defaultRateFields = {
    localExcess: 0,
  } as { localExcess?: number };

  const tgParams = {
    prefix,
    secrets: secrets ?? [],
    cors: cors ? { ...defaultCorsFields, ...cors } : defaultCorsFields,
    rate: rate ? { ...defaultRateFields, ...rate } : undefined,
  };

  core.initTypegraph({ name, dynamic, path, ...tgParams });

  const g: TypegraphBuilderArgs = {
    expose: (exports, defaultPolicy) => {
      core.expose(
        Object.entries(exports).map(([name, fn]) => [name, fn._id]),
        defaultPolicy ? getPolicyChain(defaultPolicy) : []
      );
    },
    inherit: () => {
      return new InheritDef();
    },
    rest: (graphql: string) => {
      return wit_utils.addGraphqlEndpoint(graphql);
    },
    auth: (value: Auth | RawAuth) => {
      if (value instanceof RawAuth) {
        return wit_utils.addRawAuth(value.jsonStr);
      }
      return wit_utils.addAuth(value);
    },
    ref: (name: string) => {
      return genRef(name);
    },
    configureRandomInjection: (params: { seed: number }) => {
      return core.setSeed(params.seed);
    },
    ...InjectionSource,
  };

  builder(g);

  const ret = {
    serialize(config: SerializeParams) {
      try {
        const [tgJson, ref_artifacts] = core.serializeTypegraph(
          config
        ) as Array<any>; // FIXME: bad typing?
        const result: TgFinalizationResult = {
          tgJson: tgJson,
          ref_artifacts: ref_artifacts,
        };
        return result;
      } catch (err) {
        const stack = (err as any)?.payload?.stack;
        if (stack) {
          // FIXME: jco generated code throws new Error(object) => prints [Object object]
          throw new Error(stack.join("\n"));
        }
        throw err;
      }
    },
    name,
  } as TypegraphOutput;

  if (hasCliEnv()) {
    const manager = new Manager(ret);
    await manager.run();

    // TODO solve hanging process (stdin??)
    setTimeout(() => {
      if (counter === 0) {
        log.debug("exiting");
        process.exit(0);
      }
    }, 10);
  }

  --counter;

  return ret;
}

export function genRef(name: string) {
  const value = core.refb(name, []);
  if (typeof value == "object") {
    throw new Error(JSON.stringify(value));
  }
  return new t.Typedef(value, { name });
}
